### ISSUE TITLE
Don't use safeUp() as default in migration template

### ISSUE THREAD

**Issue by tom--** [yiisoft/yii2#15212]
### What steps will reproduce the problem?

In a large data migration one might need to batch updates in order to limit transaction size.

```php
    public function safeUp()
    {
        $db = Yii::$app->db;
        $batch = 10000;
        $n = 0;
        do {
            $n += $batch;
            $db->open();
            $db
                ->createCommand(
                    "UPDATE foo
                    INNER JOIN bar ON foo.bar_id = bar.id
                    INNER JOIN baz ON baz.baz_id = baz.id
                    SET foo.blah_id = baz.blah_id
                    WHERE foo.blah_id IS NULL AND foo.id <= :n"
                )
                ->bindValue(':n', $n)
                ->execute();
        } while ($n < Foo::find()->max('id'));
    }
```

### What is the expected result?

Each command executes.

### What do you get instead?

The updates run very slowly because 10s of GB are being buffered. When `safeUp()` finally commits its transaction, mysqld tries to apply and replicate this gigantic transaction. The migration completes with a success message but the transaction was in fact rolled back because it was too big to replicate...

```
2017-11-22 16:56:11 140277977589504 [Warning] WSREP: transaction size limit (2147483647) exceeded: 2147483648
2017-11-22 16:56:11 140277977589504 [ERROR] WSREP: rbr write fail, data_len: 0, 2
```

### Additional info

It does no good to wrap each batch in an explicit transaction with `beginTransaction()` and `commit()` because `safeUp()` is wrapping the whole thing in its transaction.

In either case, `safeUp()` hides that the migration actually failed.

With MySQL, `safeUp()` is at best useless (wrapping DDL in a transaction is foolish) and in DML migrations such as this example it prevents the migration from working and masks that it failed. 

So I don't think `safeUp()` is a good default for the migration template.

| Q                | A
| ---------------- | ---
| Yii version      | v2.0.13.1
| PHP version      | 7.1.10
| Operating system | Debian 9.2

**Comment by samdark** [comment]
What do you mean by `safeUp()` hides that the migration actually failed? No error?

**Comment by schmunk42** [comment]
Related:
- https://github.com/yiisoft/yii2/issues/13831
- https://github.com/yiisoft/yii2/issues/13827
- https://github.com/yiisoft/yii2/issues/14568

**Comment by tom--** [comment]
@samdark Yes. As I said in the _What do you get instead?_ text "The migration completes with a success message but the transaction was in fact rolled back"

**Comment by samdark** [comment]
Oh... then this behavior should be fixed so transaction both rolls back and throws visible error.

**Comment by tom--** [comment]
I think that's likely to be hard because mysqld did not return an error. I imagine the transaction commits in the normal way but then Galera, which is sorta async, has to undo it when the corresponding writeset fails to replicate.

**Comment by samdark** [comment]
Sounds like MySQL cluster-specific problem...

**Comment by rawtaz** [comment]
PostgreSQL, anyone? :-)

@tom-- You said DDL in a transaction is foolish. Did you mean that in the context of MySQL or in general? I quite like that PostgreSQL actually does support DDL in transactions - that's what makes me feel that my migrations will not screw anything up in case of a failure during the migration. Just curious if you can elaborate a bit.

**Comment by samdark** [comment]
@rawtaz that was specific to MySQL cluster.

**Comment by samdark** [comment]
I don't think we can fix that from our side and since galera setups aren't that common, it's better to cover 80% of cases with common setups. @tom-- for your case you can redefine migration template.

**Comment by tom--** [comment]
@rawtaz Yes.

> With MySQL, safeUp() is at best useless (wrapping DDL in a transaction is foolish)...

I maintain the opinion that we introduce risk when we make the framework wrap migrations in transactions on MySQL as a default that's kinda hidden.

Debugging the above issue took two days until I shouted something like:

> **WHAT!** Where did this transaction come from? Oh f&*k me, what's safe up doing here?

For example, what happens if the migration mixes DDL and DML and uses safe up on MySQL? I have no idea. So it doesn't sound like a good idea to introduce the transaction behind the scenes.

**Comment by samdark** [comment]
> I have no idea.

Transaction is committed as soon as any DDL happens.

**Comment by samdark** [comment]
It's not exactly behind the scenes. It's quite explicitly named method.

**Comment by tom--** [comment]
How "behind the scenes" this is is surely debatable.

1. It's a default in a template that's not hard to overlook.
2. The method name Safe Up just sounds good. Why wouldn't I want safe? If it were Transactional Up, that might remind me of the design question I need to consider and might stand out a bit more being longer.

I think I made my point and I will shut up now.

**Comment by schmunk42** [comment]
> The method name Safe Up just sounds good. Why wouldn't I want safe? If it were Transactional Up, that might remind me of the design question I need to consider and might stand out a bit more being longer.

That's a valid point. It should be considered to rename this method (with a possible or optional fallback) in 2.1/3.0.0.

**Comment by samdark** [comment]
Renaming makes sense.

**Comment by tom--** [comment]
@rawtaz I don't think there's much controversy about that. A lot of users regard the lack of transnational DDL as a serious deficiency.
